lightning/src/ln/blinded_payment_tests.rs

```diff
@@ -2298,9 +2298,230 @@ fn test_trampoline_single_hop_receive() {
+			onion_utils::build_trampoline_onion_payloads(
+				&blinded_tail,
+				original_amt_msat,
+				&recipient_onion_fields,
+				starting_htlc_offset,
+				&None,
+			)
+			.unwrap();
+
+		if !blinded {
+			trampoline_payloads = vec![msgs::OutboundTrampolinePayload::Receive {
+				payment_data: Some(msgs::FinalOnionHopData {
+					payment_secret,
+					total_msat: original_amt_msat,
+				}),
+				sender_intended_htlc_amt_msat: original_amt_msat,
+				cltv_expiry_height: original_trampoline_cltv + starting_htlc_offset,
+			}];
+		}
+
+		let trampoline_onion_keys = onion_utils::construct_trampoline_onion_keys(
+			&secp_ctx,
+			&blinded_tail,
+			&trampoline_session_priv,
+		);
+		let trampoline_packet = onion_utils::construct_trampoline_onion_packet(
+			trampoline_payloads,
+			trampoline_onion_keys,
+			override_random_bytes,
+			&payment_hash,
+			None,
+		)
+		.unwrap();
+
+		(trampoline_packet, outer_total_msat, outer_starting_htlc_offset)
+	};
+
+	// Use a different session key to construct the replacement onion packet. Note that the
+	// sender isn't aware of this and won't be able to decode the fulfill hold times.
+	let (mut outer_payloads, _, _) = onion_utils::build_onion_payloads(
+		&route.paths[0],
+		outer_total_msat,
+		&recipient_onion_fields,
+		outer_starting_htlc_offset,
+		&None,
+		None,
+		Some(trampoline_packet),
+	)
+	.unwrap();
+	assert_eq!(outer_payloads.len(), 2);
+
+	// If we're trying to test invalid payloads, we modify Carol's *outer* onion to have values
+	// that are inconsistent with her inner onion. We need to do this manually because we
+	// (obviously) can't construct an invalid onion with LDK's built in functions.
+	match &mut outer_payloads[1] {
+		msgs::OutboundOnionPayload::TrampolineEntrypoint {
+			amt_to_forward,
+			outgoing_cltv_value,
+			..
+		} => match test_case {
+			TrampolineTestCase::Underpayment => {
+				*amt_to_forward = test_case.outer_onion_amt(original_amt_msat)
+			},
+			TrampolineTestCase::OuterCLTVLessThanTrampoline => {
+				*outgoing_cltv_value =
+					test_case.outer_onion_cltv(original_trampoline_cltv + starting_htlc_offset)
+			},
+			_ => {},
+		},
+		_ => panic!("final payload is not trampoline entrypoint"),
+	}
+
+	let outer_onion_keys =
+		onion_utils::construct_onion_keys(&secp_ctx, &route.clone().paths[0], &outer_session_priv);
+	onion_utils::construct_onion_packet(
+		outer_payloads,
+		outer_onion_keys,
+		override_random_bytes,
+		&payment_hash,
+	)
+	.unwrap()
+}
+
+// Test relay of payments to a trampoline, testing success and trampoline-related relay failures.
+// This test relies on manually replacing parts of our onion to:
+// - Test unblinded trampoline receives, which are not natively supported in LDK.
+// - To hit validation errors by manipulating the trampoline's outer packet. Without this, we would
+//   have to manually construct the onion.
+fn do_test_trampoline_relay(blinded: bool, test_case: TrampolineTestCase) {
```

Comment:
On closer look, I'm pretty fine with this test. Probably fine to add in that test coverage you mentioned? Did it seem crazy when you looked into it?

Reply:
Removed forwarding validation for a later PR, so don't need to add anything here.